### PR TITLE
feat: update jest, support mixed esm and non-esm packages

### PR DESCRIPTION
**Pull request by BRKalow**
🎟️ [Asana Task]()

---

## Description

Support an in-between state where _some_ packages use ESM, while some do not. Splits our test run into two separate commands, `test:cjs` and `test:esm`. I also went ahead and updated jest to v29, and swapped from `ts-jest` to `swc` for speed.

## PR Checklist 🚀

- [ ] Conduct thorough self-review.
- [ ] Add or update tests as appropriate.
- [ ] Write a useful description (above) to give reviewers appropriate context.
- [ ] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -1,19 +1,43 @@
...
   testEnvironment: 'node',
   verbose: true,
-  globals: {
-    'ts-jest': {
-      diagnostics: {
-        warnOnly: true,
-      },
-    },
+  transform: {
+    '^.+\\.(j|t)sx?$': ['@swc/jest'],
```
Comment (thiskevinwang) [inline jest.config.js:28]: What's the thought behind `@swc/jest`? (This is new to me)

Comment (BRKalow) [reply jest.config.js:28]: I found it much more straight forward to configure than `ts-jest` out of the box, and it's faster.

File: package.json [hunk 2]
```diff
@@ -43,8 +45,11 @@
     "lint": "./packages/cli/next-hashicorp lint",
     "release": "npm run build && changeset publish",
     "release:canary": "npm run build && changeset publish --tag canary",
-    "test": "NODE_OPTIONS=--experimental-vm-modules jest",
-    "test:watch": "jest --watch",
+    "test:esm": "TEST_ESM=true NODE_OPTIONS=--experimental-vm-modules jest",
+    "test:esm:watch": "npm run test:esm --watch",
+    "test:cjs": "jest",
+    "test:cjs:watch": "npm run test:cjs --watch",
+    "test": "npm run test:cjs && npm run test:esm",
```
Comment (thiskevinwang) [inline package.json:52]: Nice!

**Review by thiskevinwang — APPROVED** [review]
Looks reasonable to me! The whole ESM/CJS+jest world makes me 😵

**Pull request merged**